storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -347,17 +353,40 @@ func (s *peersSource) incrementalFlush(
 		shardRetriever    = shardRetrieverMgr.ShardRetriever(shard)
 		tmpCtx            = context.NewContext()
 		seriesCachePolicy = s.opts.ResultOptions().SeriesCachePolicy()
+		filePathPrefix    = s.opts.FilesystemOptions().FilesystemOptions().FilePathPrefix()
 	)
 	if seriesCachePolicy == series.CacheAllMetadata && shardRetriever == nil {
 		return fmt.Errorf("shard retriever missing for shard: %d", shard)
 	}
 
 	for start := tr.Start; start.Before(tr.End); start = start.Add(blockSize) {
-		prepared, err := flush.Prepare(nsMetadata, shard, start)
+		prepared, ok, err := flush.Prepare(nsMetadata, shard, start)
 		if err != nil {
 			return err
 		}
 
+		if !ok {
+			// If we reach this point then it means that the fileset files already
```

Comment:
Few different things:
- Deleting files unconditionally in a bootstrapper scares me. Could you include a knob in the bootsrapper options to allow users to configure whether data should be overwritten or not.
- This assumes a bunch about how we're doing bootstrapping today (always doing FS before Peers). What if we decide to run with other bootstrapping configurations in the future?

Reply:
1) I can add the knob, although I'm not sure what the value in it is...
2) I'm not sure if it really assumes that much. This code path only executes in the scenario where we've streamed all the data for a given shard/block from our peers and we're ready to write that information to disk.

One alternative would be to make the FS bootstrapper do the deletion when it detects a fileset file it can't read which is less leaky (or to archive the files instead of deleting them) but then you're deleting files for which you don't yet have a replacement which seems less good.

Also I'm open to completely alternative solutions, I just couldn't think of any

Reply:
The other alternative is asking an operator to take action, ie fail the bootstrap altogether and crash (which would start crash looping the node).

Maybe that’s ok if there was an easy way for you to decide whether to delete the bad files or not and/or stream them in with peers on next process start, etc.